p2p/discovery/broadcast.go

```diff
@@ -0,0 +1,187 @@
+	}
+	return conn, nil
+}
+
+func NewBroadcastService(ctx context.Context, peerhost host.Host, defaultPort int, interval time.Duration) (Service, error) {
+
+	// Broadcasts can only be received from 0.0.0.0 or 255.255.255.255
+	// but they can not be sent from 255.255.255.255.
+	// Therefor one listening socket is bound to 255.255.255.255 and a
+	// sending socket will be bound to each address that broadcasts
+	// should be sent from.
+	listenSock, err := bindBroadcast(net.IPv4bcast, defaultPort)
+	if err != nil {
+		return nil, err
+	}
+
+	bs := &broadcastService{
+		host:          peerhost,
+		listenSock:    listenSock,
+		broadcastPort: defaultPort,
+		active:        true,
+	}
+
+	addrs, err := getDialableListenAddrs(peerhost)
+	if err != nil {
+		bs.Close()
+		return nil, err
+	} else {
+		for _, a := range addrs {
+			if a.IP.IsGlobalUnicast() {
+				// binding to the same address and port as the TCP service
+				// embeds that address and port into broadcast messages
+				otherSock, err := bindBroadcast(a.IP, a.Port)
+				if err == nil {
+					bs.sendSocks = append(bs.sendSocks, otherSock)
+				}
+			}
+		}
+	}
+
+	// fudge the interval to dampen oscillations with other nodes
+	margin := int(interval) >> 2
+	interval += time.Duration(rand.Intn(margin))
+	interval -= time.Duration(rand.Intn(margin))
+
+	go bs.sendBroadcasts(ctx, interval)
+	go bs.recvBroadcasts(ctx)
+
+	return bs, nil
+}
+
+func (bs *broadcastService) Close() error {
+	bs.active = false
+	bs.listenSock.Close()
+	for _, conn := range bs.sendSocks {
+		conn.Close()
+	}
+	return nil
+}
+
+func (bs *broadcastService) handleBroadcast(addr *net.UDPAddr, bpeer peer.ID) {
+	// Broadcast messages are UDP, but represent IPv4 TCP sockets
+	maddr, err := manet.FromNetAddr(&net.TCPAddr{
+		IP:   addr.IP,
+		Port: addr.Port,
+	})
+	if err == nil {
+		pi := pstore.PeerInfo{
+			ID:    bpeer,
+			Addrs: []ma.Multiaddr{maddr},
+		}
+
+		bs.lk.Lock()
+		defer bs.lk.Unlock()
+		for _, n := range bs.notifees {
+			if n != nil {
+				go n.HandlePeerFound(pi)
+			}
+		}
+	}
+}
+
+func (bs *broadcastService) recvBroadcasts(ctx context.Context) {
+	var buf = make([]byte, 48)
+	for bs.active {
+		n, rAddr, err := bs.listenSock.ReadFromUDP(buf)
+		if err == nil {
+			if n != 0 {
+				rPeer, err := peer.IDB58Decode(string(buf[0:n]))
+				if err == nil && rPeer != bs.host.ID() {
+					bs.handleBroadcast(rAddr, rPeer)
+				}
+			}
+		}
+	}
+}
+
+func (bs *broadcastService) sendBroadcast() {
+	idBuf := []byte(bs.host.ID().Pretty())
+	broadcastAddr := net.UDPAddr{
+		IP:   net.IPv4bcast,
+		Port: bs.broadcastPort,
+	}
+	for _, conn := range bs.sendSocks {
+		conn.WriteToUDP(idBuf, &broadcastAddr)
+	}
+}
+
+func (bs *broadcastService) sendBroadcasts(ctx context.Context, interval time.Duration) {
+	// send an initial broadcast after a quarter interval
+	time.Sleep(interval >> 2)
+	bs.sendBroadcast()
+
+	// send broadcasts on a periodic timeout
+	ticker := time.NewTicker(interval)
+	for bs.active {
```

Comment:
Again, can't read/write to a variable from different threads without using some synchronization mechanism (lock, channel, etc.).

Really, I'd just get rid of this variable and use the context.

Reply:
It's a bool.

Reply:
> It's a bool.

I know, I'm referring to https://golang.org/ref/mem. For example, given:

```go
package main

func main() {
	var active = true

	go func() {
		active = false
	}()

	for active {
	}
}
```

Go may choose to allow the second go routine to loop forever (optimizing out the check of `active`) because no synchronization primitives were used. Try running the above code with the race detector.